RISC V build failures

I recently submitted a PR to add 4.13.0 to Opam ([here](https://github.com/ocaml/opam-repository/pull/25548)) and the build failed on RISC V platforms ([log](https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/3557c3d73b152439564e8da8d17b8c52481be422/variant/extras,riscv64-ocaml-4.14,z3.4.13.0)). This flavour:

```
# g++  -o z3  shell/datalog_frontend.o shell/dimacs_frontend.o shell/drat_frontend.o shell/main.o shell/opt_fro....
# g++ -o libz3.so -shared api/dll/dll.o api/dll/install_tactic.o api/dll/mem_initializer.o api/dll/gparam...
# /usr/bin/ld: api/api.a(api_array.o): in function `.L0 ':
# api_array.cpp:(.text+0x32): undefined reference to `__atomic_exchange_1'
# /usr/bin/ld: api_array.cpp:(.text+0x17a): undefined reference to `__atomic_exchange_1'
# /usr/bin/ld: api/api.a(api_array.o): in function `.LEHB10':
# api_array.cpp:(.text+0x320): undefined reference to `__atomic_exchange_1'
# /usr/bin/ld: api/api.a(api_array.o): in function `.L100':
# api_array.cpp:(.text+0x4e6): undefined reference to `__atomic_exchange_1'
# /usr/bin/ld: api/api.a(api_array.o): in function `.L173':
# api_array.cpp:(.text+0x6ae): undefined reference to `__atomic_exchange_1'
# /usr/bin/ld: api/api.a(api_array.o):api_array.cpp:(.text+0x910): more undefined references to `__atomic_exchange_1' follow
# collect2: error: ld returned 1 exit status
```

This is the first time Z3 is built on those platforms in the Opam CI, so it may well be a configuration problem on their end (It's Ubuntu 22.04 though, not sure what's different). But if someone with RISC V experience knows what's different on those platforms, I'd appreciate any hints!

--- a/scripts/mk_util.py
+++ b/scripts/mk_util.py
@@ -314,6 +314,23 @@ def test_fpmath(cc):
         FPMATH_FLAGS=""
         return "UNKNOWN"
 
+def test_atomic_required(cc):
+    t = TempFile('tstatomic.cpp')
+    t.add("""
+    #include <atomic>
+    std::atomic<int> x;
+    std::atomic<short> y;
+    std::atomic<char> z;
+    std::atomic<long long> w;
+    int main() {
+        ++z;
+        ++y;
+        ++w;
+        return ++x;
+    }")
+    """)
+    t.commit()
+    return exec_compiler_cmd([cc, CPPFLAGS, '', 'tstatomic.cpp', LDFLAGS, '']) != 0
 
 def find_jni_h(path):
     for root, dirs, files in os.walk(path):
@@ -555,19 +572,19 @@ def set_version(major, minor, build, revision):
             print("Set Assembly Version (BUILD):", VER_MAJOR, VER_MINOR, VER_BUILD, VER_TWEAK)
             return
 
-    # use parameters to set up version if not provided by script args            
+    # use parameters to set up version if not provided by script args
     VER_MAJOR = major
     VER_MINOR = minor
     VER_BUILD = build
     VER_TWEAK = revision
 
-    # update VER_TWEAK base on github     
+    # update VER_TWEAK base on github
     if GIT_DESCRIBE:
         branch = check_output(['git', 'rev-parse', '--abbrev-ref', 'HEAD'])
         VER_TWEAK = int(check_output(['git', 'rev-list', '--count', 'HEAD']))
-    
+
     print("Set Assembly Version (DEFAULT):", VER_MAJOR, VER_MINOR, VER_BUILD, VER_TWEAK)
-    
+
 def get_version():
     return (VER_MAJOR, VER_MINOR, VER_BUILD, VER_TWEAK)
 
@@ -1858,7 +1875,7 @@ class JavaDLLComponent(Component):
                           os.path.join('api', 'java', 'Native'))
             elif IS_OSX and IS_ARCH_ARM64:
                 out.write('\t$(SLINK) $(SLINK_OUT_FLAG)libz3java$(SO_EXT) $(SLINK_FLAGS) -arch arm64 %s$(OBJ_EXT) libz3$(SO_EXT)\n' %
-                          os.path.join('api', 'java', 'Native'))                
+                          os.path.join('api', 'java', 'Native'))
             else:
                 out.write('\t$(SLINK) $(SLINK_OUT_FLAG)libz3java$(SO_EXT) $(SLINK_FLAGS) %s$(OBJ_EXT) libz3$(SO_EXT)\n' %
                           os.path.join('api', 'java', 'Native'))
@@ -2600,6 +2617,9 @@ def mk_config():
         CXXFLAGS = '%s -fvisibility=hidden -fvisibility-inlines-hidden -c' % CXXFLAGS
         FPMATH = test_fpmath(CXX)
         CXXFLAGS = '%s %s' % (CXXFLAGS, FPMATH_FLAGS)
+        atomic_required = test_atomic_required(CXX)
+        if atomic_required:
+            LDFLAGS  = '%s -latomic' % LDFLAGS
         if LOG_SYNC:
             CXXFLAGS = '%s -DZ3_LOG_SYNC' % CXXFLAGS
         if SINGLE_THREADED:
@@ -2710,6 +2730,7 @@ def mk_config():
             print('Prefix:         %s' % PREFIX)
             print('64-bit:         %s' % is64())
             print('FP math:        %s' % FPMATH)
+            print('libatomic:      %s' % ('required' if atomic_required else 'not required'))
             print("Python pkg dir: %s" % PYTHON_PACKAGE_DIR)
             if GPROF:
                 print('gprof:          enabled')
@@ -2854,7 +2875,7 @@ def update_version():
     revision = VER_TWEAK
 
     print("UpdateVersion:", get_full_version_string(major, minor, build, revision))
-    
+
     if major is None or minor is None or build is None or revision is None:
         raise MKException("set_version(major, minor, build, revision) must be used before invoking update_version()")
     if not ONLY_MAKEFILES:
@@ -3068,7 +3089,7 @@ def mk_bindings(api_files):
           z3py_output_dir=get_z3py_dir(),
           dotnet_output_dir=dotnet_output_dir,
           java_input_dir=java_input_dir,
-          java_output_dir=java_output_dir,                                  
+          java_output_dir=java_output_dir,
           java_package_name=java_package_name,
           ml_output_dir=ml_output_dir,
           ml_src_dir=ml_output_dir

--- a/scripts/mk_util.py
+++ b/scripts/mk_util.py
@@ -327,7 +327,7 @@ def test_atomic_required(cc):
         ++y;
         ++w;
         return ++x;
-    }")
+    }
     """)
     t.commit()
     fails_without = exec_compiler_cmd([cc, CPPFLAGS, '', 'tstatomic.cpp', LDFLAGS, '']) != 0

--- a/scripts/mk_util.py
+++ b/scripts/mk_util.py
@@ -330,7 +330,10 @@ def test_atomic_required(cc):
     }")
     """)
     t.commit()
-    return exec_compiler_cmd([cc, CPPFLAGS, '', 'tstatomic.cpp', LDFLAGS, '']) != 0
+    fails_without = exec_compiler_cmd([cc, CPPFLAGS, '', 'tstatomic.cpp', LDFLAGS, '']) != 0
+    ok_with = exec_compiler_cmd([cc, CPPFLAGS, '', 'tstatomic.cpp', LDFLAGS + ' -latomic', '']) == 0
+    return fails_without and ok_with
+
 
 def find_jni_h(path):
     for root, dirs, files in os.walk(path):

